### PR TITLE
Add SE2 UKF-M example

### PR DESCRIPTION
**Pull request by artivis**
Add a SE2 localization example based on the paper ['A Code for Unscented Kalman Filtering on Manifolds (UKF-M)'](https://arxiv.org/pdf/2002.00878.pdf), M. BROSSARD, A. BARRAU and S. BONNABEL.

### CODE REVIEW

File: examples/se2_localization.cpp [hunk 1]
```diff
@@ -280,9 +280,11 @@ int main()
...
 
         // DEBUG
-        cout << "X simulated : " << X_simulation.log().coeffs().transpose() << endl;
-        cout << "X estimated : " << X.log().coeffs().transpose() << endl;
-        cout << "X unfilterd : " << X_unfiltered.log().coeffs().transpose() << endl;
+        cout << "X simulated : " << X_simulation.log().coeffs().transpose() << "." << endl;
+        cout << "X estimated : " << X.log().coeffs().transpose()
+             << ". |d|=" << (X_simulation - X).squaredWeightedNorm() <<endl;
+        cout << "X unfilterd : " << X_unfiltered.log().coeffs().transpose()
+             << ". |d|=" << (X_simulation - X_unfiltered).squaredWeightedNorm() <<endl;
```
Comment (joansola) [inline examples/se2_localization.cpp:287]: why squared weighted norm? If you name it `|d|` should be the norm, not squared norm

File: examples/se2_localization_ukfm.cpp [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+ *  This file is:
+ *  (c) 2018 Joan Sola @ IRI-CSIC, Barcelona, Catalonia
+ *
+ *  This file is part of `manif`, a C++ template-only library
+ *  for Lie theory targeted at estimation for robotics.
+ *  Manif is:
+ *  (c) 2018 Jeremie Deray @ IRI-UPC, Barcelona
+ *  ---------------------------------------------------------
+ *
+ *  ---------------------------------------------------------
```
Comment (joansola) [inline examples/se2_localization_ukfm.cpp:17]: Fix authorship and doc details throughout this comments section

File: examples/se2_localization_ukfm.cpp [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+    Matrix<double, Rp, 2*DoF> yj;
+    Array2d                   y_sigmas;
+    Matrix2d                  R;
+    std::vector<Vector2d>     measurements(landmarks.size());
+
+    y_sigmas << 0.01, 0.01;
+    R        = (y_sigmas * y_sigmas).matrix().asDiagonal();
+
+
+    // Declare UFK variables
```
Comment (joansola) [inline examples/se2_localization_ukfm.cpp:218]: UKF?

**Review by joansola — CHANGES_REQUESTED** [review]
I am unfamiliar with the implementation of the UKF. I added some silly comments to be adtressed. I'll check more things but submit this review in the meanwhile.

**Comment by artivis** [comment]
@joansola This example runs but the results are not quite satisfying. Would you please help me figure where (if) I made a mistake?? I've been following and reviewing again and again the paper mentioned above. I've also peeked at the [paper's associated online code](https://github.com/CAOR-MINES-ParisTech/ukfm) without luck so far. I'll try implementing this example with their framework see if I get any different results...

These are the results of the new UKF-M-based example:
<details><summary>CLICK ME</summary>
<p>

```
X STATE     :    X      Y    THETA
----------------------------------
X initial   : +0.0000 +0.0000 +0.0000
----------------------------------
X simulated : +0.1000 +0.0000 +0.0500.
X estimated : +0.0958 -0.1080 +0.1054. |d|=+0.0172
X unfilterd : +0.1680 -0.0211 +0.1066. |d|=+0.0114
----------------------------------
X simulated : +0.2000 -0.0000 +0.1000.
X estimated : +0.1895 -0.1557 +0.1797. |d|=+0.0346
X unfilterd : +0.2783 -0.0261 +0.1824. |d|=+0.0200
----------------------------------
X simulated : +0.3000 -0.0000 +0.1500.
X estimated : +0.3020 +0.0374 +0.1333. |d|=+0.0018
X unfilterd : +0.3049 +0.0043 +0.1357. |d|=+0.0004
----------------------------------
X simulated : +0.4000 -0.0000 +0.2000.
X estimated : +0.4117 +0.1272 +0.1283. |d|=+0.0232
X unfilterd : +0.3198 +0.1064 +0.1293. |d|=+0.0248
----------------------------------
X simulated : +0.5000 +0.0000 +0.2500.
X estimated : +0.5196 +0.1945 +0.1388. |d|=+0.0530
X unfilterd : +0.4128 +0.2058 +0.1378. |d|=+0.0641
----------------------------------
X simulated : +0.6000 -0.0000 +0.3000.
X estimated : +0.6124 +0.0450 +0.2732. |d|=+0.0030
X unfilterd : +0.4928 +0.2446 +0.2716. |d|=+0.0683
----------------------------------
X simulated : +0.7000 -0.0000 +0.3500.
X estimated : +0.7394 +0.1814 +0.2379. |d|=+0.0471
X unfilterd : +0.6687 +0.2477 +0.2344. |d|=+0.0703
----------------------------------
X simulated : +0.8000 -0.0000 +0.4000.
X estimated : +0.8382 +0.1608 +0.2976. |d|=+0.0370
X unfilterd : +0.8516 +0.2019 +0.2924. |d|=+0.0513
----------------------------------
X simulated : +0.9000 -0.0000 +0.4500.
X estimated : +0.9047 +0.0235 +0.4341. |d|=+0.0008
X unfilterd : +0.9414 +0.2430 +0.4286. |d|=+0.0563
----------------------------------
X simulated : +1.0000 +0.0000 +0.5000.
X estimated : +0.9738 -0.0887 +0.5591. |d|=+0.0112
X unfilterd : +0.9952 +0.1795 +0.5546. |d|=+0.0477
----------------------------------
```

</p>
</details>

In comparison, here are the results of the existing `se2_localization` example:
<details><summary>CLICK ME</summary>
<p>

```
X STATE     :    X      Y    THETA
----------------------------------
X initial   : +0.0000 +0.0000 +0.0000
----------------------------------
X simulated : +0.1000 +0.0000 +0.0500.
X estimated : +0.1000 +0.0080 +0.0453. |d|=+0.0001
X unfilterd : +0.1680 -0.0211 +0.1066. |d|=+0.0114
----------------------------------
X simulated : +0.2000 -0.0000 +0.1000.
X estimated : +0.1971 -0.0102 +0.1030. |d|=+0.0001
X unfilterd : +0.2783 -0.0261 +0.1824. |d|=+0.0200
----------------------------------
X simulated : +0.3000 -0.0000 +0.1500.
X estimated : +0.3025 +0.0009 +0.1524. |d|=+0.0000
X unfilterd : +0.3049 +0.0043 +0.1357. |d|=+0.0004
----------------------------------
X simulated : +0.4000 -0.0000 +0.2000.
X estimated : +0.3983 +0.0004 +0.1986. |d|=+0.0000
X unfilterd : +0.3198 +0.1064 +0.1293. |d|=+0.0248
----------------------------------
X simulated : +0.5000 +0.0000 +0.2500.
X estimated : +0.4937 +0.0058 +0.2476. |d|=+0.0001
X unfilterd : +0.4128 +0.2058 +0.1378. |d|=+0.0641
----------------------------------
X simulated : +0.6000 -0.0000 +0.3000.
X estimated : +0.6054 -0.0096 +0.3056. |d|=+0.0002
X unfilterd : +0.4928 +0.2446 +0.2716. |d|=+0.0683
----------------------------------
X simulated : +0.7000 -0.0000 +0.3500.
X estimated : +0.7078 +0.0124 +0.3427. |d|=+0.0003
X unfilterd : +0.6687 +0.2477 +0.2344. |d|=+0.0703
----------------------------------
X simulated : +0.8000 -0.0000 +0.4000.
X estimated : +0.8017 +0.0040 +0.3976. |d|=+0.0000
X unfilterd : +0.8516 +0.2019 +0.2924. |d|=+0.0513
----------------------------------
X simulated : +0.9000 -0.0000 +0.4500.
X estimated : +0.8988 -0.0067 +0.4537. |d|=+0.0001
X unfilterd : +0.9414 +0.2430 +0.4286. |d|=+0.0563
----------------------------------
X simulated : +1.0000 +0.0000 +0.5000.
X estimated : +1.0031 +0.0008 +0.4985. |d|=+0.0000
X unfilterd : +0.9952 +0.1795 +0.5546. |d|=+0.0477
```

</p>
</details>

File: examples/se2_localization_ukfm.cpp (hunk 2 above)
Comment (artivis) [reply examples/se2_localization_ukfm.cpp:17]: I'll fix all the doc once the code itself is :+1:

**Comment by joansola** [comment]
Yea it seems UKF is not doing the job...

**Comment by joansola** [comment]
Checkout this compiling error (MacOSX)

```terminal
[ 77%] Building CXX object test/se3/CMakeFiles/gtest_se3.dir/gtest_se3.cpp.o
In file included from /Users/jsola/dev/manif/test/se3/gtest_se3.cpp:3:
In file included from /Users/jsola/dev/manif/include/manif/SE3.h:9:
/Users/jsola/dev/manif/include/manif/impl/se3/SE3_base.h:254:40: error: 'norm' following the 'template' keyword
      does not refer to a template
  MANIF_ASSERT(abs(quaternion.template norm()-Scalar(1)) <
                                       ^
/Users/jsola/dev/manif/include/manif/impl/se3/SE3_base.h:265:3: note: in instantiation of function template
      specialization 'manif::SE3Base<manif::SE3<double> >::quat<Eigen::Matrix<double, 4, 1, 0, 4, 1> >'
      requested here
  quat(so3.coeffs());
  ^
/Users/jsola/dev/manif/test/se3/gtest_se3.cpp:115:7: note: in instantiation of member function
      'manif::SE3Base<manif::SE3<double> >::quat' requested here
  se3.quat(axisAngle);
      ^
/usr/local/include/eigen3/Eigen/src/Core/MatrixBase.h:193:34: note: declared as a non-template here
    EIGEN_DEVICE_FUNC RealScalar norm() const;
                                 ^
1 error generated.
```

**Comment by artivis** [comment]
> Checkout this compiling error (MacOSX)

Erf, what's your compiler now? I'll see to add it to CI

**Comment by joansola** [comment]
vermicelle:build jsola$ clang++ --version
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.7.0

**Comment by joansola** [comment]
So you used the matlab code in https://github.com/CAOR-MINES-ParisTech/ukfm/tree/master/matlab as the reference code?

**Comment by artivis** [comment]
> So you used the matlab code in https://github.com/CAOR-MINES-ParisTech/ukfm/tree/master/matlab as the reference code?

Yes both the Matlab and Python implementations.

**Comment by joansola** [comment]
I was having a look but got lost in the process. Tried the following without success:

- Use `llt.matrixU()` instead of `matrixL()`

and, also, later I saw that it is in fact `matrixL()` what we need (in case it makes any difference).

**Comment by artivis** [comment]
@joansola this demo is finally working! I figured from the paper that the main issue was the use of the right-minus instead of the left-minus (see [here](https://github.com/artivis/manif/blob/eb7b509136f9aabf258f5f234d4579c05b3979c8/examples/se2_localization_ukfm.cpp#L317) and [here](https://github.com/artivis/manif/blob/eb7b509136f9aabf258f5f234d4579c05b3979c8/examples/se2_localization_ukfm.cpp#L334)). The error is of the same magnitude as EFK when compared head to head, not sure which algo is better in average on our small example tho.
Anyway, this PR is ready for review :+1: .

**Review by joansola — APPROVED** [review]
se2_localization_ukfm has good error levels like the original se2_localization. Approving

**Pull request merged**